### ISSUE TITLE
batched initial state input influence numerical precision

### ISSUE THREAD

**Issue by Jooyuza** [TuringQ/deepquantum#29]
When running `\tests\test_init_state_batch.py`,  pytest sometimes fail to `allclose` the batched and unbatched result, even when absolute tolerence `atol=1e-4`. This indicates the _same_ initial state in batched and unbatched mode will generate _different_ amplitude and probability value for some final states. 

i.e., tensor([-0.1382-0.0771j]) for unbatched initial state and tensor([-0.1477-0.0825j]) for batched ones.

![image](https://github.com/user-attachments/assets/6baa77f4-54cd-480a-9467-27dc404cd46e)

When looking into a concret example, set `init_state = [2,1,1]` and `init_state = [[2,1,1],[2,2,2]]`, with the same circuit parameters `data = torch.tensor([-0.9713, -0.0058, -0.0766, -0.0955, -1.0791,  1.0020,  0.3544, -0.0251])`.
Then the probabilties to observe `final_state = [2,0,2]`, getting from the forward function, are `0.3248` and `0.3712` respectively.
![image](https://github.com/user-attachments/assets/8556b4bc-a706-4799-b353-a82b24baa5d1)
If there's no expand mode, then results are the same.

**Comment by Jooyuza** [comment]
Setting a higher cutoff for expand mode doesn't work. `cutoff = 7`
![image](https://github.com/user-attachments/assets/b1e5925d-4f4e-40ab-8995-46c250f62aac)

**Comment by Jooyuza** [comment]
For such case, the probality of final state [2,2,0] is wrong as well. In fact, the probabilities of batched result are not normalised.
![image](https://github.com/user-attachments/assets/5ced6a3d-53ff-4cd1-b1ff-7d476ec9a66c)

**Comment by sansiro77** [comment]
This bug comes from `product_factorial`. For example, `torch.exp(torch.lgamma(torch.tensor(5).double() + 1)).long()` is `tensor(119)`.
However, `torch.exp(torch.lgamma(torch.tensor(5).double() + 1)).float()` or `torch.exp(torch.lgamma(torch.tensor(5).float() + 1)).long()` is fine.

**Comment by sansiro77** [comment]
Solved in #27
